assets/src/edit-story/components/library/panes/text/textSets/textSets.js

```diff
@@ -83,7 +82,34 @@ function TextSets() {
   );
 
   useEffect(() => {
-    getTextSets().then(setTextSets);
+    window.clearTimeout(loadingTimeoutRef.current);
+    isLoadingRef.current = false;
+
+    setFilteredTextSets(selectedCat ? textSets[selectedCat] : allTextSets);
+    setRenderedTextSets([]);
+  }, [selectedCat, textSets, allTextSets]);
+
+  useEffect(() => {
+    if (isLoadingRef.current) {
+      return;
+    }
+
+    if (renderedTextSets.length < filteredTextSets.length) {
+      isLoadingRef.current = true;
+
+      loadingTimeoutRef.current = window.setTimeout(() => {
+        isLoadingRef.current = false;
+        setRenderedTextSets([
+          ...renderedTextSets,
+          filteredTextSets[renderedTextSets.length],
+        ]);
+      }, 10);
+    }
+  }, [renderedTextSets, filteredTextSets]);
```

Comment:
I think we can get rid of the refs if we do something like this:
```
  const numRendered = renderedTextSets.length;
  useEffect(() => {
    if (numRendered >= filteredTextSets.length) {
      return () => {};
    }

    const loadingTimeoutId = window.setTimeout(() => {
      setRenderedTextSets((rendered) => [
        ...rendered,
        filteredTextSets[rendered.length],
      ]);
    }, 10);

    return () => {
      window.clearTimeout(loadingTimeoutId);
    };
  }, [numRendered, filteredTextSets]);
```

Should also make the timeout logic specific to this effect.

Reply:
Would also make it easy to swap out with `requestIdleCallback` if we wanted to go down that route

Reply:
Thanks for the refactor!  Works like a charm.  When I was first writing this out, my intention was to not use any refs but then I ran into a bug that was causing the useEffect to trigger more than it should, so I put in the loading ref and then when I actually fixed the bug, things smoothed out but I didn't back out of the decision.  Thanks for the second eyes, this is much cleaner!!